### PR TITLE
features/309-nft-tests-rework

### PR DESCRIPTION
**Pull request by simongalasso**

### CODE REVIEW

File: nfts/src/tests/extrinsics.rs [hunk 1]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
+mod create {
 	use super::*;
 
 	#[test]
-	fn cannot_transfer_delegated_nfts() {
-		ExtBuilder::new_build(vec![(ALICE, 100)]).execute_with(|| {
-			let nft_id = NFTs::create_nft(ALICE, vec![0], None).unwrap();
-			assert_ok!(NFTs::set_viewer(nft_id, Some(BOB)));
+	fn create_ok_with_no_series() {
+		ExtBuilder::default().caps(vec![(ALICE, 1000)]).build().execute_with(|| {
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:21]: Use `new_build` (check out test `cannot_transfer_delegated_nfts`) instead of deafults + caps + build combo

File: nfts/src/tests/extrinsics.rs [hunk 2]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
-mod transfer {
+mod create {
 	use super::*;
 
 	#[test]
-	fn cannot_transfer_delegated_nfts() {
-		ExtBuilder::new_build(vec![(ALICE, 100)]).execute_with(|| {
-			let nft_id = NFTs::create_nft(ALICE, vec![0], None).unwrap();
-			assert_ok!(NFTs::set_viewer(nft_id, Some(BOB)));
+	fn create_ok_with_no_series() {
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:20]: You should have two happy tests.

One with series id set to None and one with a series id set to a new value.

Comment (markopoloparadox) [reply nfts/src/tests/extrinsics.rs:20]: Also, the test name should not contain the word "ok" in them.

File: nfts/src/tests/extrinsics.rs [hunk 3]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
+		ExtBuilder::default().caps(vec![(ALICE, 1000)]).build().execute_with(|| {
+			// State checks
+			assert_eq!(NFTs::nft_id_generator(), 0);
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:23]: You don't need that.

File: nfts/src/tests/extrinsics.rs [hunk 4]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
+			assert_eq!(NFTs::series_id_generator(), 0);
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:24]: You don't need that.

File: nfts/src/tests/extrinsics.rs [hunk 5]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
+			// Create NFT with new serie id while there is no series already registered
+			assert_ok!(NFTs::create(
+				alice.clone(),
+				data.ipfs_reference.clone(),
+				Some(data.series_id.clone()),
+			));
+			let nft_id = NFTs::nft_id_generator() - 1;
+
+			// Events checks
+			assert_eq!(
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:42]: First we check the storage and then the event

File: nfts/src/tests/extrinsics.rs [hunk 6]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
+				System::events().last().unwrap().event,
+				Event::NFTs(NFTsEvent::NFTCreated {
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:44]: This is ugly. Move it outside just like it was done in the auction tets

File: nfts/src/tests/extrinsics.rs [hunk 7]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
+			let serie_id = vec![50];
+			let ipfs_reference = vec![1];
+			let data = NFTData::new_default(ALICE, ipfs_reference.clone(), serie_id.clone());
+			let alice_balance = Balances::free_balance(ALICE);
 
-			let ok = NFTs::transfer(origin(ALICE), nft_id, BOB);
-			assert_noop!(ok, Error::<Test>::CannotTransferDelegatedNFTs);
+			// Create NFT with new serie id while there is no series already registered
+			assert_ok!(NFTs::create(
+				alice.clone(),
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:35]: use the origin() function from auction tests.

File: nfts/src/tests/extrinsics.rs [hunk 8]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
+		ExtBuilder::default().caps(vec![(ALICE, 1000)]).build().execute_with(|| {
+			// State checks
+			assert_eq!(NFTs::nft_id_generator(), 0);
+			assert_eq!(NFTs::series_id_generator(), 0);
+
+			// Initial state
+			let alice: mock::Origin = origin(ALICE);
+			let serie_id = vec![50];
+			let ipfs_reference = vec![1];
+			let data = NFTData::new_default(ALICE, ipfs_reference.clone(), serie_id.clone());
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:30]: You can directly instantiate  series id an ipfs reference here. There is no need to create separate variables for them.

File: nfts/src/tests/extrinsics.rs [hunk 9]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
-		ExtBuilder::new_build(vec![(ALICE, 100)]).execute_with(|| {
-			let nft_id = NFTs::create_nft(ALICE, vec![0], None).unwrap();
-			assert_ok!(NFTs::set_viewer(nft_id, Some(BOB)));
+	fn create_ok_with_no_series() {
+		ExtBuilder::default().caps(vec![(ALICE, 1000)]).build().execute_with(|| {
+			// State checks
+			assert_eq!(NFTs::nft_id_generator(), 0);
+			assert_eq!(NFTs::series_id_generator(), 0);
+
+			// Initial state
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:26]: Unnecessary comment.

File: nfts/src/tests/extrinsics.rs [hunk 10]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
+	fn create_ok_associated_with_existing_serie() {
+		ExtBuilder::default().caps(vec![(ALICE, 1000)]).build().execute_with(|| {
+			// State checks
+			assert_eq!(NFTs::nft_id_generator(), 0);
+			assert_eq!(NFTs::series_id_generator(), 0);
+
+			// Initial state
+			let alice: mock::Origin = origin(ALICE);
+			let serie_id = vec![50];
+			let ipfs_1_reference = vec![1];
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:72]: Unnecessary. Instantiate the ipfs directly where it is needed.

File: nfts/src/tests/extrinsics.rs [hunk 11]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
+			let ipfs_2_reference = vec![2];
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:73]: Unnecessary. Instantiate the ipfs directly where it is needed.

File: nfts/src/tests/extrinsics.rs [hunk 12]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
+	#[test]
+	fn create_ok_associated_with_existing_serie() {
+		ExtBuilder::default().caps(vec![(ALICE, 1000)]).build().execute_with(|| {
+			// State checks
+			assert_eq!(NFTs::nft_id_generator(), 0);
+			assert_eq!(NFTs::series_id_generator(), 0);
+
+			// Initial state
+			let alice: mock::Origin = origin(ALICE);
+			let serie_id = vec![50];
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:71]: 'series'

File: nfts/src/tests/extrinsics.rs [hunk 13]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
-mod burn {
-	use super::*;
+	#[test]
+	fn create_ok_associated_with_existing_serie() {
+		ExtBuilder::default().caps(vec![(ALICE, 1000)]).build().execute_with(|| {
+			// State checks
+			assert_eq!(NFTs::nft_id_generator(), 0);
+			assert_eq!(NFTs::series_id_generator(), 0);
+
+			// Initial state
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:69]: Unnecessary. Please remove this in all the places where we have it.

File: nfts/src/tests/extrinsics.rs [hunk 14]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
 	}
-}
 
-mod burn {
-	use super::*;
+	#[test]
+	fn create_ok_associated_with_existing_serie() {
+		ExtBuilder::default().caps(vec![(ALICE, 1000)]).build().execute_with(|| {
+			// State checks
+			assert_eq!(NFTs::nft_id_generator(), 0);
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:66]: Unnecessary. Please remove this in all the places where we have it.

File: nfts/src/tests/extrinsics.rs [hunk 15]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
+			assert_eq!(NFTs::series_id_generator(), 0);
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:67]: Unnecessary. Please remove this in all the places where we have it.

File: nfts/src/tests/extrinsics.rs [hunk 16]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
 		})
 	}
-}
 
-mod burn {
-	use super::*;
+	#[test]
+	fn create_ok_associated_with_existing_serie() {
+		ExtBuilder::default().caps(vec![(ALICE, 1000)]).build().execute_with(|| {
+			// State checks
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:65]: Unnecessary. Please remove this in all the places where we have it.

File: nfts/src/tests/extrinsics.rs [hunk 17]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
+			assert_eq!(NFTs::data(0), Some(data.clone()));
+			assert_eq!(Balances::free_balance(ALICE), alice_balance - NFTs::nft_mint_fee());
 		})
 	}
-}
 
-mod burn {
-	use super::*;
+	#[test]
+	fn create_ok_associated_with_existing_serie() {
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:63]: Use new_build (check out test cannot_transfer_delegated_nfts) instead of deafults + caps + build combo
Please do this in all tests.

File: nfts/src/tests/extrinsics.rs [hunk 18]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
+			let alice: mock::Origin = origin(ALICE);
+			let serie_id = vec![50];
+			let ipfs_reference = vec![1];
+			let data = NFTData::new_default(ALICE, ipfs_reference.clone(), serie_id.clone());
+			let alice_balance = Balances::free_balance(ALICE);
 
-			let ok = NFTs::transfer(origin(ALICE), nft_id, BOB);
-			assert_noop!(ok, Error::<Test>::CannotTransferDelegatedNFTs);
+			// Create NFT with new serie id while there is no series already registered
+			assert_ok!(NFTs::create(
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:34]: Move the NFTs::create call return value into a separate variable first. Do this for all the tests that have multi-lined assert_ok!() for extrinsic calls.

File: nfts/src/tests/extrinsics.rs [hunk 19]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
+					owner: data_2.owner,
+					series_id: data_2.series_id.clone(),
+					ipfs_reference: ipfs_2_reference,
+					mint_fee: NFTs::nft_mint_fee(),
+				})
+			);
+
+			// Final state checks
+			assert_eq!(NFTs::series_id_generator(), 0);
+			assert_eq!(NFTs::series(&data_1.series_id), Some(NFTSeriesDetails::new(ALICE, true)));
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:126]: Move the instantiation of "NFTSeriesDetails::new(ALICE, true)" to a variable. Do the same thing for the line below

File: nfts/src/tests/extrinsics.rs (hunk 5 above)
Comment (markopoloparadox) [reply nfts/src/tests/extrinsics.rs:42]: Do this for all other tests too.

File: nfts/src/tests/extrinsics.rs [hunk 20]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
+			assert_eq!(Balances::free_balance(ALICE), alice_balance - NFTs::nft_mint_fee() * 2);
+		})
+	}
 
 	#[test]
-	fn cannot_burn_delegated_nfts() {
-		ExtBuilder::new_build(vec![(ALICE, 100)]).execute_with(|| {
-			let nft_id = NFTs::create_nft(ALICE, vec![0], None).unwrap();
-			assert_ok!(NFTs::set_viewer(nft_id, Some(BOB)));
+	fn create_error_too_short_name() {
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:135]: The name of the test should be exactly the same as the name of the error. In this case the test should be called "ipfs_reference_is_too_short"

Comment (markopoloparadox) [reply nfts/src/tests/extrinsics.rs:135]: Please do this for all other tests

File: nfts/src/tests/extrinsics.rs [hunk 21]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
+			assert_eq!(NFTs::nft_id_generator(), 0);
+			assert_eq!(NFTs::series_id_generator(), 0);
 
-			let ok = NFTs::burn(origin(ALICE), nft_id);
-			assert_noop!(ok, Error::<Test>::CannotBurnDelegatedNFTs);
+			// Initial state
+			let alice: mock::Origin = origin(ALICE);
+
+			// Should fail and storage should remain empty
+			let ok = NFTs::create(alice.clone(), vec![], None);
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:145]: If you need mock::Origin in multiple places, than it's OK to have it first inside a variable. But in cases like here where you only use it once it's OK to just call the "origin()" function.

File: nfts/src/tests/extrinsics.rs [hunk 22]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
+			// State checks
+			assert_eq!(NFTs::nft_id_generator(), 0);
+			assert_eq!(NFTs::series_id_generator(), 0);
+
+			// Initial state
+			let alice: mock::Origin = origin(ALICE);
+
+			let series_id = Some(vec![51]);
+			assert_ok!(NFTs::create(alice.clone(), vec![50], series_id.clone()));
+			NFTs::finish_series(alice.clone(), series_id.clone().unwrap()).unwrap();
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:219]: user asserts and not unwraps

File: nfts/src/tests/extrinsics.rs [hunk 23]
```diff
@@ -9,375 +9,665 @@ fn origin(account: u64) -> mock::Origin {
...
 
 	#[test]
-	fn cannot_delegate_nfts_listed_for_sale() {
+	fn transfer_error_uncompleted_serie() {
 		ExtBuilder::new_build(vec![(ALICE, 100)]).execute_with(|| {
-			let nft_id = NFTs::create_nft(ALICE, vec![0], None).unwrap();
-			assert_ok!(NFTs::set_listed_for_sale(nft_id, true));
+			// Initial state
+			let alice: mock::Origin = origin(ALICE);
+			let nft_id = <NFTs as NFTTrait>::create_nft(ALICE, vec![0], None).unwrap();
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:284]: You are constantly creating the same nft in all other tests. This is a good case for having this NFT already in the genesis config

File: nfts/src/tests/extrinsics.rs (hunk 1 above)
Comment (simongalasso) [reply nfts/src/tests/extrinsics.rs:21]: Done

File: nfts/src/tests/extrinsics.rs (hunk 4 above)
Comment (simongalasso) [reply nfts/src/tests/extrinsics.rs:24]: Done, it was already there before rework, I did not know why so I let it, now removed.

File: nfts/src/tests/extrinsics.rs (hunk 5 above)
Comment (simongalasso) [reply nfts/src/tests/extrinsics.rs:42]: Done

File: nfts/src/tests/extrinsics.rs (hunk 6 above)
Comment (simongalasso) [reply nfts/src/tests/extrinsics.rs:44]: Done

File: nfts/src/tests/extrinsics.rs (hunk 7 above)
Comment (simongalasso) [reply nfts/src/tests/extrinsics.rs:35]: Done, I used it above, but it was only used once so I replaced alice.clone() by origin(ALICE).

File: nfts/src/tests/extrinsics.rs (hunk 8 above)
Comment (simongalasso) [reply nfts/src/tests/extrinsics.rs:30]: Done

File: nfts/src/tests/extrinsics.rs (hunk 9 above)
Comment (simongalasso) [reply nfts/src/tests/extrinsics.rs:26]: Done

File: nfts/src/tests/extrinsics.rs (hunk 12 above)
Comment (simongalasso) [reply nfts/src/tests/extrinsics.rs:71]: Done

File: nfts/src/tests/extrinsics.rs (hunk 13 above)
Comment (simongalasso) [reply nfts/src/tests/extrinsics.rs:69]: Done

**Comment by markopoloparadox** [comment]
@simongalasso It seems that you are pushing your changes into a wrong branch.

File: nfts/src/tests/extrinsics.rs (hunk 14 above)
Comment (simongalasso) [reply nfts/src/tests/extrinsics.rs]: Done

File: nfts/src/tests/extrinsics.rs (hunk 15 above)
Comment (simongalasso) [reply nfts/src/tests/extrinsics.rs]: Done

File: nfts/src/tests/extrinsics.rs (hunk 16 above)
Comment (simongalasso) [reply nfts/src/tests/extrinsics.rs]: Done

File: nfts/src/tests/extrinsics.rs (hunk 17 above)
Comment (simongalasso) [reply nfts/src/tests/extrinsics.rs:55]: Done

File: nfts/src/tests/extrinsics.rs (hunk 18 above)
Comment (simongalasso) [reply nfts/src/tests/extrinsics.rs]: Done, In Auction tests, sometimes it's made like you say sometimes like I did.

File: nfts/src/tests/extrinsics.rs (hunk 20 above)
Comment (simongalasso) [reply nfts/src/tests/extrinsics.rs]: Done

File: nfts/src/tests/extrinsics.rs (hunk 21 above)
Comment (simongalasso) [reply nfts/src/tests/extrinsics.rs:125]: Done

File: nfts/src/tests/extrinsics.rs (hunk 23 above)
Comment (simongalasso) [reply nfts/src/tests/extrinsics.rs:198]: Sometimes tests don't need nft, ex: NFTNotFound, so with this I would instantiate it for nothing, I could do special builders for recurrent cases ?

File: nfts/src/tests/extrinsics.rs (hunk 2 above)
Comment (simongalasso) [reply nfts/src/tests/extrinsics.rs]: Done

File: nfts/src/tests/extrinsics.rs [hunk 24]
```diff
@@ -9,375 +9,548 @@ fn origin(account: u64) -> mock::Origin {
...
+				alice.clone(),
+				data.ipfs_reference.clone(),
+				Some(data.series_id.clone()),
+			);
+			assert_ok!(ok);
+			let nft_id = NFTs::nft_id_generator() - 1;
+
+			// Final state checks
+			assert_eq!(NFTs::series_id_generator(), 0);
+			let nft_series_details = Some(NFTSeriesDetails::new(ALICE, true));
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:37]: Put it above all the assert_eq (It should be before line 36)

File: nfts/src/tests/extrinsics.rs [hunk 25]
```diff
@@ -9,375 +9,548 @@ fn origin(account: u64) -> mock::Origin {
...
+			let ipfs_reference = vec![1];
+			let alice_balance = Balances::free_balance(ALICE);
+
+			// Create NFT with new serie id while there is no series already registered
+			let ok = NFTs::create(origin(ALICE), ipfs_reference.clone(), None);
+			assert_ok!(ok);
+			let nft_id = NFTs::nft_id_generator() - 1;
+
+			// Final state checks
+			let data = NFTs::data(0);
```
Comment (markopoloparadox) [inline nfts/src/tests/extrinsics.rs:68]: Magic number 0

Comment (markopoloparadox) [reply nfts/src/tests/extrinsics.rs:68]: Just call unwrap here and call it a day.

Comment (markopoloparadox) [reply nfts/src/tests/extrinsics.rs:68]: Actually. Call assert!(data.is_some())
And then resign data = data.unwrap();

Comment (markopoloparadox) [reply nfts/src/tests/extrinsics.rs:68]: You know, just call unwrap and that's all

File: nfts/src/tests/extrinsics.rs (hunk 23 above)
Comment (markopoloparadox) [reply nfts/src/tests/extrinsics.rs:198]: The tests that don't need it will not use it. Quite simple.

File: nfts/src/tests/extrinsics.rs (hunk 25 above)
Comment (simongalasso) [reply nfts/src/tests/extrinsics.rs]: I used nft_id, and is_some(), Done

File: nfts/src/tests/extrinsics.rs (hunk 24 above)
Comment (simongalasso) [reply nfts/src/tests/extrinsics.rs:36]: Done

File: nfts/src/tests/extrinsics.rs (hunk 23 above)
Comment (simongalasso) [reply nfts/src/tests/extrinsics.rs]: Done

File: nfts/src/lib.rs [hunk 26]
```diff
@@ -512,16 +512,6 @@ impl<T: Config> traits::NFTTrait for Pallet<T> {
 
 		Ok(())
 	}
-
-	fn set_viewer(id: NFTId, value: Option<Self::AccountId>) -> DispatchResult {
```
Comment (markopoloparadox) [inline nfts/src/lib.rs:516]: Don't remove stuff that are unrelated to this task :)

**Review by markopoloparadox — APPROVED** [review]
